exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/columnreaders/NullableColumnReader.java

```diff
@@ -29,117 +29,152 @@
-          }
-          definitionLevelsRead = 0;
+          || (definitionLevelsRead >= pageReader.currentPageCount)) {
+        if (!pageReader.next()) {
+          break;
         }
-        lastValueWasNull = true;
-        runLength = 0;
-        if (lastRunBrokenByNull ) {
-          nullsFound = 1;
-          lastRunBrokenByNull = false;
-        } else  {
-          nullsFound = 0;
-        }
-        // loop to find the longest run of defined values available, can be preceded by several nulls
-        while(indexInOutputVector < recordsToReadInThisPass
-            && indexInOutputVector < valueVec.getValueCapacity()
-          && definitionLevelsRead < pageReader.currentPageCount) {
+        //New page. Reset the definition level.
+        currentDefinitionLevel = -1;
+        definitionLevelsRead = 0;
+        recordsReadInThisIteration = 0;
+      }
+
+      nullRunLength = 0;
+      runLength = 0;
+
+      //
+      // Let's skip the next run of nulls if any ...
+      //
+
+      // If we are reentering this loop, the currentDefinitionLevel has already been read
+      if (currentDefinitionLevel < 0) {
+        currentDefinitionLevel = pageReader.definitionLevels.readInteger();
+      }
+      haveMoreData = readCount < recordsToReadInThisPass
+          && writeCount + nullRunLength < valueVec.getValueCapacity()
+          && definitionLevelsRead < pageReader.currentPageCount;
+      while (haveMoreData && currentDefinitionLevel < columnDescriptor
+          .getMaxDefinitionLevel()) {
+        readCount++;
+        nullRunLength++;
+        definitionLevelsRead++;
+        haveMoreData = readCount < recordsToReadInThisPass
+            && writeCount + nullRunLength < valueVec.getValueCapacity()
+            && definitionLevelsRead < pageReader.currentPageCount;
+        if (haveMoreData) {
           currentDefinitionLevel = pageReader.definitionLevels.readInteger();
-          definitionLevelsRead++;
-          indexInOutputVector++;
-          totalDefinitionLevelsRead++;
-          if ( currentDefinitionLevel < columnDescriptor.getMaxDefinitionLevel()){
-            // a run of non-null values was found, break out of this loop to do a read in the outer loop
-            if ( ! lastValueWasNull ){
-              lastRunBrokenByNull = true;
-              break;
-            }
-            nullsFound++;
-            lastValueWasNull = true;
-          }
-          else{
-            if (lastValueWasNull){
-              runLength = 0;
-              lastValueWasNull = false;
-            }
-            runLength++;
-            castedVectorMutator.setIndexDefined(indexInOutputVector - 1);
-          }
         }
-        valuesReadInCurrentPass += nullsFound;
+      }
+      //
+      // Write the nulls if any
+      //
+      if (nullRunLength > 0) {
+        int writerIndex =
+            ((BaseDataValueVector) valueVec).getBuffer().writerIndex();
+        castedBaseVector.getBuffer().setIndex(0, writerIndex + (int) Math
```

Comment:
Yes. This code was useful when the Bit data type was a single bit. It is no longer so and this check is not needed.